### ISSUE TITLE
`/members` peer handler is not guarded with linearizable check

### ISSUE THREAD

**Issue by chaochn47** [etcd-io/etcd#16687]
### Bug report criteria

- [X] This bug report is not security related, security issues should be disclosed privately via security@etcd.io.
- [X] This is not a support request, support requests should be raised in the etcd [discussion forums](https://github.com/etcd-io/etcd/discussions).
- [X] You have read the etcd [bug reporting guidelines](https://github.com/etcd-io/etcd/blob/main/Documentation/contributor-guide/reporting_bugs.md).
- [X] Existing open issues along with etcd [frequently asked questions](https://etcd.io/docs/latest/faq) have been checked and this is not a duplicate.

### What happened?

Restart etcd server right after member reconfiguration and query the member list via HTTP /members, [its handler](https://github.com/etcd-io/etcd/blob/release-3.4/etcdserver/api/etcdhttp/peer.go#L107) will bypass the linearizable check. The member list response would be stale during bootstrap (restart) where the members are restored from v2 store and WAL is still replaying...

### What did you expect to happen?

When peer certificate is not used, `/members` should be protected by linearizable check. 

### How can we reproduce it (as minimally and precisely as possible)?

1. checkout code in https://github.com/chaochn47/etcd/tree/v3.4.20-eks.0-reproduce-member-name-mismatch
3. `cd integration && go test -v -run TestReproduceMemberNameMismatch`

### Anything else we need to know?

Relevant to discussion 

> > Restart etcd server right after member reconfiguration and query the member list via HTTP `/members`, [its handler](https://github.com/etcd-io/etcd/blob/release-3.4/etcdserver/api/etcdhttp/peer.go#L107) will bypass the linearizable check.
> 
> Actually the main also has this "issue", but this might not be a problem. Two reasons:
> 
>     * The `/members` is only supposed to accessed during bootstrap by a new etcd member who has just been added to the cluster;
> 
>     * The `/members` is protected by peer certificate; in other words, users shouldn't be able to access this endpoint at all.
> 
> 
> EDIT: but I still think it may be better to guard the `/members` by linearizable check. Please raise a separate issue to track it, and we can discuss it separately.

https://github.com/etcd-io/etcd/issues/16666#issuecomment-1744601073

### Etcd version (please run commands below)

<details>

```console
$ etcd --version
# paste output here

$ etcdctl version
# paste output here
```

</details>

### Etcd configuration (command line flags or environment variables)

<details>

# paste your configuration here

</details>

### Etcd debug information (please run commands below, feel free to obfuscate the IP address or FQDN in the output)

<details>

```console
$ etcdctl member list -w table
# paste output here

$ etcdctl --endpoints=<member list> endpoint status -w table
# paste output here
```

</details>

**Comment by serathius** [comment]
QQ, why we have a non-grpc endpoint for listing members? Can someone do some archaeology because for me it seems like legacy from v2 API.

**Comment by chaochn47** [comment]
From what I can tell, it is used in runtime reconfiguration (adding a new member). 

https://github.com/etcd-io/etcd/blob/0f4d7a7483127b78b6560ae4cc6509ade5fb1162/server/etcdserver/bootstrap.go#L289.

When the new member process starts, it needs to verify the existing cluster ID matches the local `--initial-cluster` configuration, assign with this new member ID, validate the max learners does not exceed the configuration... etc.

**Comment by serathius** [comment]
This looks like code used to bootstrap member joining a existing cluster. For example when increasing cluster size. From brief look, call to members is used to get mapping between member names and ids. 

I would need to double check how fetching of those ids are later synchronized with raft, but I suspect that getting the freshest data might not be necessary as long as raft is properly initiated and replayed.

As this is internal endpoint our only worry should be whether this stale data here could impact member bootstrap. We should be very careful when adding a linearizability requirement here. It could bring more harm than benefit. First I would want to see that we are able to find and reproduce issue with etcd member bootstrap that could be caused by stale data returned on `/members` endpoint.

**Comment by chaochn47** [comment]
> It could bring more harm than benefit. 

Agree. Here is a historical issue I opened https://github.com/etcd-io/etcd/issues/14174. The conclusion is https://github.com/etcd-io/etcd/issues/14174#issuecomment-1190909437

> Discussed in #14175, the feasible solution in short term is
> 
>     * Add retry in client side to ensure the membership reconfiguration applied to all members.
> 
>     * Wait until retry succeeds, start new member.
